text/0119-dynamic-component-creation.md

```diff
@@ -0,0 +1,412 @@
+
+For context, the current design requires a component author define an arbitrary custom element name which will be used as the name for all constructors passed to the `lwc:dynamic` directive, creating a 1:N mapping.
+
+For example:
+
+```html
+<template>
+    <button onclick={loadCtor}>Click to change constructor</button>
+    <x-lazy lwc:dynamic={customCtor}></x-lazy>
+</template>
+```
+
+```javascript
+import { LightningElement } from "lwc";
+import About from 'c/about';
+import Home from 'c/home';
+
+export default class extends LightningElement {
+    customCtor;
+    ctors = [About, Home];
+
+    loadCtor() {
+        if (this.customCtor) {
+            // Append to the end of the array
+            this.ctors.push(this.customCtor);
+        }
+
+        // Remove the first element in the array
+        this.customCtor = this.ctors.shift();
+    }
+}
+```
+
+Will always render the same tag regardless of the constructor.
+
+```html
+<button onclick={loadCtor}>Click to change constructor</button>
+<x-lazy>
+    <!-- c-custom-constructor content -->
+</x-lazy>
+```
+
+This proposal aims to overcome the issues with the current design by resolving the custom element name at compile time and using it as the name of the dynamic component that is created at runtime.
+
+## Prior art
+
+Several other frameworks have the concept of dynamically rendering components:
+
+- [Svelte: `<svelte:component this={expression}>`](https://svelte.dev/docs#template-syntax-svelte-component)
+- [Vue: `<component :is="...">`](https://vuejs.org/guide/essentials/component-basics.html#dynamic-components)
+- [Angular: viewContainerRef](https://angular.io/guide/dynamic-component-loader)
+- [Lit: static expressions](https://lit.dev/docs/templates/expressions/#static-expressions)
+- [Aura: $A.createComponent](https://developer.salesforce.com/docs/atlas.en-us.lightning.meta/lightning/js_cb_dynamic_cmp_async.htm)
+
+Frameworks that use JSX such as [React](https://reactjs.org/docs/rendering-elements.html#rendering-an-element-into-the-dom) and [Stencil](https://stenciljs.com/docs/templating-jsx) can dynamically render components directly into the markup using expressions.
+
+The goal of managing dynamic components in LWC is to follow the same component lifecycle as all other LWC custom elements.  Having said that, taking a similar approach as Svelte and Vue seems the most natural and intuitive path forward.
+
+## Detailed design
+
+A new internal `Map` called `componentRegisteredNameMap` is introduced to store the custom element name and an internal API `getComponentRegisteredName` is introduced to retrieve the custom element name.
```

Comment:
I decided to switch to storing the custom element name on a `Map` because we decided that the [custom element's name should not be observable to the component authors](https://github.com/salesforce/lwc-rfcs/pull/71#discussion_r1062982660).

The main appeal to me of storing the name on the `LightningElementConstructor` was to be able to use it with `this.template.querySelector` but since we can use `lwc:ref` instead, it doesn't seem necessary anymore.